### ISSUE TITLE
zero learning rate when resume training

### ISSUE THREAD

**Issue by sysu-zjw** [facebookresearch/DPR#99]
Hi, I just finished training DPR with 50 epochs on my data and tried to continually train another 50 epochs. Below is my training command. When resume training, I change `num_train_epochs 50` to `num_train_epochs 100`. 

```
python -m torch.distributed.launch \
	--nproc_per_node=8 train_dense_encoder.py \
	--max_grad_norm 2.0 \
	--encoder_model_type hf_bert \
	--pretrained_model_cfg bert-base-uncased \
	--seed 12345 \
	--sequence_length 256 \
	--warmup_steps 1237 \
	--batch_size 6 \
	--do_lower_case \
	--train_file "/export/data/jzhoubu/workspace/DPR/data/data/retriever/train.json" \
	--dev_file "/export/data/jzhoubu/workspace/DPR/data/data/retriever/dev.json" \
	--output_dir $OUT_DIR \
	--learning_rate 2e-05 \
	--num_train_epochs 100 \
	--dev_batch_size 16 \
	--shuffle_positive_ctx \
	--val_av_rank_start_epoch 30 |& tee $OUT_DIR/log
```

After resuming training (i.e. from epoch 50 to epoch 100), the learning rate is 0.0000 in every epoch and the reported `Av.rank validation` also remains the same across different epochs. Below is part of the log:

```
Initialized host lccpu16.cse.ust.hk as d.rank 5 on device=cuda:5, n_gpu=1, world size=8
16-bits training: False 
 **************** CONFIGURATION **************** 
adam_betas                     -->   (0.9, 0.999)
adam_eps                       -->   1e-08
batch_size                     -->   6
checkpoint_file_name           -->   dpr_biencoder
dev_batch_size                 -->   16
dev_file                       -->   /export/data/jzhoubu/workspace/DPR/data/data/retriever/dev-h1.json
device                         -->   cuda:5
distributed_world_size         -->   8
do_lower_case                  -->   True
dropout                        -->   0.1
encoder_model_type             -->   hf_bert
eval_per_epoch                 -->   1
fix_ctx_encoder                -->   False
fp16                           -->   False
fp16_opt_level                 -->   O1
global_loss_buf_sz             -->   150000
gradient_accumulation_steps    -->   1
hard_negatives                 -->   1
learning_rate                  -->   2e-05
local_rank                     -->   5
log_batch_step                 -->   100
max_grad_norm                  -->   2.0
model_file                     -->   None
n_gpu                          -->   1
no_cuda                        -->   False
num_train_epochs               -->   100.0
other_negatives                -->   0
output_dir                     -->   /export/data/jzhoubu/workspace/DPR/output/dense_retriever.h1.se.bs6.e100
pretrained_file                -->   None
pretrained_model_cfg           -->   bert-base-uncased
projection_dim                 -->   0
seed                           -->   12345
sequence_length                -->   256
shared_encoder                 -->   True
shuffle_positive_ctx           -->   True
train_file                     -->   /export/data/jzhoubu/workspace/DPR/data/data/retriever/train.json
train_files_upsample_rates     -->   None
train_rolling_loss_step        -->   100
val_av_rank_bsz                -->   128
val_av_rank_hard_neg           -->   30
val_av_rank_max_qs             -->   10000
val_av_rank_other_neg          -->   30
val_av_rank_start_epoch        -->   30
warmup_steps                   -->   0
weight_decay                   -->   0.0
 **************** CONFIGURATION **************** 
***** Initializing components for training *****
Checkpoint files ['/export/data/jzhoubu/workspace/DPR/output/dense_retriever.h1.se.bs6.e100/dpr_biencoder.49.1885']
Reading saved model from /export/data/jzhoubu/workspace/DPR/output/dense_retriever.h1.se.bs6.e100/dpr_biencoder.49.1885
```

```
***** Epoch 50 *****
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Epoch: 50: Step: 1/1885, loss=0.000062, lr=0.000000
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Train batch 100
Avg. loss per last 100 batches: 0.001281
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
Epoch: 50: Step: 101/1885, loss=0.025711, lr=0.000000
```

Does anybody have ideas on why this happens when I resume training from a saved checkpoint?

**Comment by vlad-karpukhin** [comment]
Hi @sysu-zjw ,
I noticed the same problem a while ago. It might be related with some pytorch updates. Learning rate scheduler state now can't load its saved state and continue training with correct learning rate.
It doesn't throw any exception or warning.
We have a fix but it is in a separate branch as of now.
See how we changed get_schedule_linear() method logic. (https://github.com/facebookresearch/DPR/blob/multi_task_training/dpr/utils/model_utils.py#L111)

I plan to integrate that fix into a new release of DPR soon.

**Comment by vlad-karpukhin** [comment]
Should be fixed with the new release. Closing this one.

**Comment by mug2mag** [comment]
@vlad-karpukhin 
hi
I have cloned the newest release code but also have the same question. The `loss` and `lr` are all almost 0. Can you help to explain the reason? Very appreciated.

**Comment by nikoletta-toth** [comment]
Hi!

Same here.
I load in the checkpoint and resume the training with my own dataset, and the learning rate is 0.
However, when I resume the training with the provided nq dataset, the learning rate loaded in looks fine.

**Comment by vlad-karpukhin** [comment]
this might be related to the checkpoint's learning rate schedule being loaded from saved state and thinking it should already be zero upon resuming.
Can you try to disable lr scheduler loading by setting ignore_checkpoint_optimizer=true as a command line argument?
